build/tasks/AnalyzeBuildGraph.cs

```diff
@@ -0,0 +1,221 @@
+
+        [Required]
+        public ITaskItem[] Artifacts { get; set; }
+
+        // Artifacts that already shipped from repos
+        [Required]
+        public ITaskItem[] ShippedArtifacts { get; set; }
+
+        [Required]
+        public string Properties { get; set; }
+
+        public string StartGraphAt { get; set; }
+
+        /// <summary>
+        /// The order in which to build repositories
+        /// </summary>
+        [Output]
+        public ITaskItem[] RepositoryBuildOrder { get; set; }
+
+        public void Cancel()
+        {
+            _cts.Cancel();
+        }
+
+        public override bool Execute()
+        {
+            var packageArtifacts = Artifacts.Select(ArtifactInfo.Parse)
+                .OfType<ArtifactInfo.Package>()
+                .Where(p => !p.IsSymbolsArtifact);
+
+            var factory = new SolutionInfoFactory(Log, BuildEngine5);
+            var props = MSBuildListSplitter.GetNamedProperties(Properties);
+
+            Log.LogMessage(MessageImportance.High, $"Beginning cross-repo analysis on {Solutions.Length} solutions. Hang tight...");
+
+            var solutions = factory.Create(Solutions, props, _cts.Token);
+            Log.LogMessage($"Found {solutions.Count} and {solutions.Sum(p => p.Projects.Count)} projects");
+
+            if (_cts.IsCancellationRequested)
+            {
+                return false;
+            }
+
+            EnsureConsistentGraph(packageArtifacts, solutions);
+            RepositoryBuildOrder = GetRepositoryBuildOrder(packageArtifacts, solutions.Where(s => s.ShouldBuild));
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private struct VersionMismatch
+        {
+            public SolutionInfo Solution;
+            public ProjectInfo Project;
+            public string PackageId;
+            public string ActualVersion;
+            public NuGetVersion ExpectedVersion;
+        }
+
+        private void EnsureConsistentGraph(IEnumerable<ArtifactInfo.Package> packages, IEnumerable<SolutionInfo> solutions)
+        {
+             var shippedPackageMap = ShippedArtifacts
+                .Select(ArtifactInfo.Parse)
+                .OfType<ArtifactInfo.Package>()
+                .Where(p => !p.IsSymbolsArtifact)
+                .ToDictionary(p => p.PackageInfo.Id, p => p, StringComparer.OrdinalIgnoreCase);
+
+            // ensure versions cascade
+            var buildPackageMap = packages.ToDictionary(p => p.PackageInfo.Id, p => p, StringComparer.OrdinalIgnoreCase);
+
+            var inconsistentVersions = new List<VersionMismatch>();
+            var reposThatShouldPatch = new HashSet<string>();
+
+            // holy crap, o^4
+            foreach (var sln in solutions)
+            foreach (var proj in sln.Projects)
+            foreach (var tfm in proj.Frameworks)
+            foreach (var dependency in tfm.Dependencies)
+            {
+                if (!buildPackageMap.TryGetValue(dependency.Key, out var package)) continue;
```

Comment:
When does this happen?

Reply:
I'll add a comment. This checks that the package is one that is actually being built by universe, and not an external or already-shipped dependency.

Reply:
Hmm we need to probably maintain a list of external dependencies as well. I'd imagine we want to know if a corefx package we depended on updated and whether we need to bump our versions.

Reply:
It's like you read my mind. The next thing I wanted to do was add this so we can also issue errors/warnings about undefined external dependencies